2.11.x fix problems with addition of elements and dropping

sageserpent-open opened the pull request:
This branch fixes the current 67 test failures seen on the upstream branch point.

This includes issue #6.

I've left the work unmerged for now, as I'm sure there are opportunities to refactor these fixes into more elegant / performant code - but they do get the tests to pass. I've added an additional application and a worksheet that I used to reproduce and debug individual test failures, plus issues I saw originally when calling 'take' and 'drop' on mutable tree bags.

There are a couple of corrections for typos mixed in.

Commit:
Added a worksheet showing some obvious failures with '+=' and 'drop/dropWhile', also added another console application for the sake of simple debugging.

Commit:
Correct typos.

Commit:
Rather ugly fixes that correct 35 of the 67 test failures prior to this commit.

Commit:
Now down to 24 test failures.

Commit:
Tidying up.

Commit:
All tests pass now - mutable tree bag was not correctly wrapping the underlying immutable  tree bag due to a bug in mapping over of the bucket iterator element.

nicolasstucki commented on scratchpad.sc:
```diff
@@ -0,0 +1,27 @@
+ttb drop 7
+
+
+ttb dropWhile (_ < 7)
```
You should not add this file to the commit.

nicolasstucki replied:
Maybe create a test with the contents of this file.

nicolasstucki commented on src/main/scala/QuickTest.scala:
```diff
@@ -0,0 +1,18 @@
+import scala.collection.mutable
+
+
+
```
Could you remove the double blank lines, leave maximum one between the statement.

nicolasstucki commented:
You should also squash all the changes that fixed something and use as commit message "Fix ..." and another single commit.

nicolasstucki commented:
That is all.

sageserpent-open commented:
The commits address several different bugs - the purpose of that work was simply to get the 67 tests failing upstream to pass, hence the use of several commits. Unfortunately I don't seem to be able to create issues on my fork, which is why I haven't used the 'fixes/closes' style comments (although that can be done for the existing issue #6). I suggest the best way to go is for you to take the branch and do whatever form of partial / squashed merge suits you on your side, should you wish to proceed further with it.

nicolasstucki commented:
Lets, leave it simple. Just remove the worksheet and i'll merge. The changes in the code LGTM.

Commit:
Remove worksheet as part of review of pull request #14.

sageserpent-open commented:
Sounds good to me, thanks - I'll write a test to cover taking and dropping to replace the worksheet. Leave it with me for a bit before you merge...

Commit:
Add test to reproduce 'splitAt', 'take' and 'drop' bugs a spart of pull request #14 review.

Commit:
Merge in additional bug reproduction test for 'splitAt', 'drop' and 'take'. Test now passes after merging with newer implementation.

sageserpent-open commented:
OK, I've pushed commits 6036257, cb7618c and c070540 (merge of the two previous commits).

There is a new test '_it should "split into parts which when recombined yield the original"_' in 'nonEmptyBagBehaviour'.

Let me know if you need a new pull request raised.

nicolasstucki commented on ignore.conf:
```diff
@@ -0,0 +1,4 @@
+target
+libraries
+.idea
+project
```
The `project` should not be ignored.

Commit:
Consider nested SBT project artifacts to be version-controlled, without bringing in the rest of the cruft generated as part of the build. Addresses review of pull request #14.

sageserpent-open commented:
No, I disagree with you.

But you have a point: it _is_ correct to include the artifacts for plugins or nested SBT projects - so I've added in nested inclusion rules for Plastic SCM (which I use). That way I don't worry about the cruft in _any_ of the potential projects, be they at any level of nesting.

It's not a showstopper for me if you want to delete this file when merging - the Git ignore file is unchanged on the branch in question, and I can tell Plastic SCM to use rules elsewhere to avoid polluting my pending changes view. Your call...

sageserpent-open commented:
Changes to 'ignore.conf' with inclusion rules pushed out in 75247bd.

nicolasstucki commented:
@sageserpent-open thanks for the fixes.

sageserpent-open commented:
You're very welcome, Nicolas - I've been using TreeBag in one of my own private projects; wanted to give back as I'd known about the dropping issue for a while. Thanks for sorting out the pull request so promptly.

nicolasstucki commented:
Appart from #6 is there another issue from the list that got solved with this PR?

sageserpent-open commented:
No, the other fixes were all addressing test failures rather than existing issues.

sageserpent-open commented:
Hi Nicolas - I was looking at another issue that was highlighted in PR #20 - when I updated my GitHub fork to pull your latest changes, I noticed that the commits  6036257, cb7618c and c070540 mentioned above never got merged - they addressed some of your review comments above in this PR - in particular, you wanted a more idiomatic ScalaTest test that isolated the problem addressed by  this PR.

Assuming that you want the additional commits, would you like to re-merge from this branch, or do you want me to cherry-pick those commits on to another pull-request branch?
